### PR TITLE
Report: use community detector lib, print similar files info

### PR DESCRIPTION
**Pull request by carlosms**
Part of #60, this is a continuation of #90.

In `report.py` the community detection output is written to a new parquet file.
`ReportApp.scala` calls this python command and reads the new parquet file. With the communities of IDs it goes to the DB to fetch info about each file and prints it.

There is a new `scripts/install_python.sh` in order to run the report command in Travis CI.

For this PR the output is pretty simple:
```
No duplicates found.

2 similar files:
	https://github.com/whatupdave/pil/blob/f829a9c2c8d41733c3363e61d8ad9ebdae8072a1/PIL/ImageStat.py
	https://github.com/python-pillow/Pillow/blob/c013188594ba44fcd42e1adca3e4744e3983c2a0/src/PIL/ImageStat.py

2 similar files:
	https://github.com/python-pillow/Pillow/blob/c013188594ba44fcd42e1adca3e4744e3983c2a0/Tests/test_file_fitsstub.py
	https://github.com/python-pillow/Pillow/blob/c013188594ba44fcd42e1adca3e4744e3983c2a0/Tests/test_file_hdf5stub.py
```

### CODE REVIEW

File: src/main/scala/tech/sourced/gemini/ReportApp.scala [hunk 1]
```diff
@@ -69,13 +74,31 @@ object ReportApp extends App {
...
+
+        cassandra.close()
+        cluster.close()
 
-      print(report)
+        print(report)
+        println()
+        printCommunities(reportCommunities)
+      } else {
+        log.error(s"Failed to execute '${pythonCmd}'")
```
[the remark this answers is not in the document]
Comment (carlosms) [reply src/main/scala/tech/sourced/gemini/ReportApp.scala]: It is not straightforward to capture both the command exit code and the output. But as the code is right now any python output to stdout/stderr is forwarded to the report app stdout.

So if something fails you will see the problem, and then the exact command that was used. Like this:
```
Traceback (most recent call last):
  File "src/main/python/community-detector/report.py", line 3, in <module>
    import numpy
ModuleNotFoundError: No module named 'numpy'
ERROR 16:27:01 tech.sourced.gemini.ReportApp$ (ReportApp.scala:95) - Failed to execute 'python3 src/main/python/community-detector/report.py src/main/python/community-detector/parquets/'
```

File: src/main/scala/tech/sourced/gemini/Gemini.scala [hunk 2]
```diff
@@ -494,5 +502,36 @@ object Gemini {
...
+    val idToSha1 = for ((elem, id) <- elementIds) yield (id, elem.split("@")(1))
+
+    // Transform communities of element IDs to communities of sha1s
+    // Equivalent to apollo graph.py BatchedCommunityResolver._gen_hashes
+    // https://github.com/src-d/apollo/blob/f51c5a92c24cbedd54b9b30bab02f03e51fd27b3/apollo/graph.py#L295
+    val communitiesSha1 = communities.map {
+      case (communityId, community) => community.collect {
+        case id if (idToSha1.contains(id)) => idToSha1(id)
+      }
+    }.filter(_.size > 1)
```
Comment (bzz) [inline src/main/scala/tech/sourced/gemini/Gemini.scala]: It took me a while to parse nested patten matching.

Would this be equivalent of

```scala
    val communitiesSha1 = communities
      .map { case (_, community) =>
        community
          .filter(idToSha1.contains)
          .map(idToSha1)
      }
      .filter(_.size > 1)
```

?

File: src/main/scala/tech/sourced/gemini/ReportApp.scala [hunk 3]
```diff
@@ -166,4 +201,32 @@ object ReportApp extends App {
...
+    val reader = AvroParquetReader.builder[GenericRecord](parquetFilePath).build()
+
+    var result = mutable.Map[Int, List[Int]]()
+
+    var nextRecord: GenericRecord = null
+
+    while ( {
+      nextRecord = reader.read
+      nextRecord
+    } != null) {
```
Comment (bzz) [inline src/main/scala/tech/sourced/gemini/ReportApp.scala]: Wow, this looks like written in C.

Could you find any other options to re-write this iteration construct?

Comment (carlosms) [reply src/main/scala/tech/sourced/gemini/ReportApp.scala]: I tried to in c24eab1, other suggestions are welcome.

File: src/main/scala/tech/sourced/gemini/ReportApp.scala [hunk 4]
```diff
@@ -69,13 +74,31 @@ object ReportApp extends App {
         case `groupByMode` => ReportExpandedGroup(gemini.reportCassandraGroupBy(cassandra))
       }
 
-      val (connectedComponents, elsToBuckets) = gemini.findConnectedComponents(cassandra)
+      val (connectedComponents, elsToBuckets, elementIds) = gemini.findConnectedComponents(cassandra)
       saveConnectedComponents(connectedComponents, elsToBuckets, config.ccDirPath)
 
-      cassandra.close()
-      cluster.close()
+      val pythonCmd = s"python3 src/main/python/community-detector/report.py ${config.ccDirPath}"
```
Comment (bzz) [inline src/main/scala/tech/sourced/gemini/ReportApp.scala]: I suggest a very simple refactoring of this method, so it's logic is more transparent and maintainable in a long run:

```scala
// cassandra & gemini setup

val duplicateReport = ...
val similarityReport = ... // what is called `reportCommunities` now

print(duplicateReport)
printCommunities(similarityReport)
// cassandra teardown
```

File: src/main/scala/tech/sourced/gemini/ReportApp.scala [hunk 5]
```diff
@@ -166,4 +200,25 @@ object ReportApp extends App {
...
   }
 
+  def readCommunities(dirPath: String): List[(Int, List[Int])] = {
+    val parquetFilePath = new Path(s"$dirPath/communities.parquet")
+    val reader = AvroParquetReader.builder[GenericRecord](parquetFilePath).build()
+
+    Iterator
+      .continually(reader.read)
+      .takeWhile(_ != null)
+      .map(record => {
```
Comment (bzz) [inline src/main/scala/tech/sourced/gemini/ReportApp.scala]: Very minor, but usually in such cases just

```scala
.map { record => 
  // do stuff
}
.toList
```

is enough.

Comment (smacker) [reply src/main/scala/tech/sourced/gemini/ReportApp.scala]: reader doesn't have `map`. I saw this pattern with `Iterator.continually.takeWhile` in engine and jgit a lot. So I think we can keep it as is.

Comment (bzz) [reply src/main/scala/tech/sourced/gemini/ReportApp.scala]: :+1: there is just extra `()` that could be avoided here

**Pull request merged**